editor/src/messages/portfolio/document/utility_types/network_interface/resolved_types.rs

```diff
@@ -0,0 +1,305 @@
+			NodeInput::Import { import_index, .. } => {
+				// Get the input type of the encapsulating node input
+				let Some((encapsulating_node, encapsulating_path)) = network_path.split_last() else {
+					return TypeSource::Error("Could not get type of import in document network");
+				};
+				self.input_type(&InputConnector::node(*encapsulating_node, *import_index), encapsulating_path)
+			}
+			NodeInput::Scope(_) => TypeSource::Compiled(concrete!(())),
+			NodeInput::Reflection(document_node_metadata) => TypeSource::Compiled(document_node_metadata.ty()),
+			NodeInput::Inline(_) => TypeSource::Compiled(concrete!(InlineRust)),
+		}
+	}
+
+	// Gets the default tagged value for an input. If its not compiled, then it tries to get a valid type. If there are no valid types, then it picks a random implementation
+	pub fn tagged_value_from_input(&mut self, input_connector: &InputConnector, network_path: &[NodeId]) -> TaggedValue {
+		let guaranteed_type = match self.input_type(input_connector, network_path) {
+			TypeSource::Compiled(compiled) => compiled,
+			TypeSource::TaggedValue(value) => value,
+			TypeSource::DocumentNodeDefinition(definition) => definition,
+			TypeSource::Unknown => {
+				let mut valid_types = self.valid_input_types(input_connector, network_path);
+
+				match valid_types.pop() {
+					Some(valid_type) => valid_type,
+					None => {
+						match self.random_type_for_connector(input_connector, network_path) {
+							Some(random_type) => random_type,
+							// If there are no connected protonodes then we give up and return the empty type
+							None => concrete!(()),
+						}
+					}
+				}
+			}
+			TypeSource::Error(e) => {
+				log::error!("Error getting tagged_value_from_input for {input_connector:?} {e}");
+				concrete!(())
+			}
+		};
+		TaggedValue::from_type_or_none(&guaranteed_type)
+	}
+
+	pub fn valid_input_types(&mut self, input_connector: &InputConnector, network_path: &[NodeId]) -> Vec<Type> {
+		let InputConnector::Node { node_id, input_index } = input_connector else {
+			// An export can have any type connected to it
+			return vec![graph_craft::generic!(T)];
+		};
+		let Some(implementation) = self.implementation(node_id, network_path) else {
+			log::error!("Could not get node implementation in valid_input_types");
+			return Vec::new();
+		};
+		match implementation {
+			DocumentNodeImplementation::Network(_) => {
+				let nested_path = [network_path, &[*node_id]].concat();
+				let Some(outward_wires) = self.outward_wires(&nested_path) else {
+					log::error!("Could not get outward wires in valid_input_types");
+					return Vec::new();
+				};
+				let Some(inputs_from_import) = outward_wires.get(&OutputConnector::Import(*input_index)) else {
+					log::error!("Could not get inputs from import in valid_input_types");
+					return Vec::new();
+				};
+
+				let intersection: HashSet<Type> = inputs_from_import
+					.clone()
+					.iter()
```

Comment:
I don't think having a clone() followed by an `.iter()` makes sense, could you just remove the clone?

Reply:
Its required since .valid_input_types requires a mutable borrow of self